### PR TITLE
Add ImageNet-D

### PR DESCRIPTION
**Pull request by georgepachitariu**
Hi, this is the first draft. 
I only started with "Clipart" subset. Please have a look.

Things left to do in this PR:

- [x] 1. Add the other subsets of Im-D;
- [x] 2. Run it all once;

### CODE REVIEW

File: examples/imagenet_d_helper/map_classes.py [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+
+    pred_label_visda_tensor = mapping_vector[pred_label].long()
+
+    return pred_label_visda_tensor
+
+
+def get_ambiguous_classes(matching_names):
+
+    # these are the ambiguous classes
+    ambiguous_classes = [
```
Comment (kotekjedi) [inline examples/imagenet_d_helper/map_classes.py:390]: This variable is never used. Can we make a dictionary from it and assign classes for `ambiguous_matching ` in some kind of for-loop?

File: examples/imagenet_d_helper/map_classes.py [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+        if not os.path.exists(target_folder_class):
+            os.makedirs(target_folder_class)
+        try:
+            allFiles_path_jpg = visda_location + "/" + folder + "/*.jpg"
+            allFiles_path_png = visda_location + "/" + folder + "/*.png"
+            allFiles_jpg = glob.glob(allFiles_path_jpg)
+            allFiles_png = glob.glob(allFiles_path_png)
+            allFiles = allFiles_jpg + allFiles_png
+            for file in allFiles:
+                newFile = target_folder_class + "/" + file.split("/")[-1]
```
Comment (kotekjedi) [inline examples/imagenet_d_helper/map_classes.py:77]: I am having a problem with this line: 

`Cannot find the path '...\\aircraft_carrier\\painting_001_000001.jpg' -> '...\\visda_symlinks\\painting\\aircraft_carrier/aircraft_carrier\\painting_001_000001.jpg'`

I think the problem caused by `file.split("/")[-1]` and backward slash.

My suggestion is to change it to smth like this:
```suggestion
                newFile = os.path.join(target_folder_class, os.path.normpath(os.path.basename(file)))
```

File: examples/imagenet_d_helper/map_classes.py [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+    return map_dict
+
+
+def get_imagenet_visda_mapping(visda_dir, map_dict):
+    # visda2imagenet
+    matching_names = dict()
+    matching_labels = dict()
+    map_dict_visda = dict()
+
+    label = 0
```
[the remark this answers is not in the document]
Comment (georgepachitariu) [reply examples/imagenet_d_helper/map_classes.py:24]: True, yes. Thanks for catching it!

File: examples/imagenet_d_helper/map_classes.py [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+        except FileExistsError:
+            pass
+
+    # final mapping and cleaning with the symlinks
+    matching_names, matching_labels = get_imagenet_visda_mapping(
+        symlinks_location, map_dict
+    )
+
+    mapping_vector = torch.zeros((1000))
+    if torch.cuda.is_available():
```
[the remark this answers is not in the document]
Comment (kotekjedi) [reply examples/imagenet_d_helper/map_classes.py:89]: For example, I have a GPU on my PC, but it is small, and I can't evaluate ResNet-50 with it.

So, for ResNet-18, I use the "cpu" option, and for ResNet-50, I use "cuda"; this function will return an array allocated on "cuda," and I need to update the code manually to put it back on "cpu."

I think it is better to explicitly control which device you have an array and make it dependent on the model's device type.

**Review by stes — DISMISSED** [review]
Hi @georgepachitariu @EvgeniaAR  , thanks a lot for your contribution. We are working towards the first release of the benchmark and for this, released our [contribution guide](https://github.com/shift-happens-benchmark/icml-2022/blob/main/CONTRIBUTING.rst) with the aim of unifying the implementation of all contributions.

Please have a look, the required changes should be relatively small and are mostly concerns with placing the files within the package, and adding documentation.

From reviewing your PR, the main required changes are:

- [ ] Place the files into their own package inside the `tasks/` folder (@georgepachitariu )
- [ ] Add a README and optionally citation info (@EvgeniaAR )
- [ ] We are also looking into the best ways to perform unit or integration tests and would be happy to get some feedback on how this would be most effective for you benchmark. Would it be possible to add a "dry-run/demo" option to the benchmark, and e.g. run it only on a small subset of the data? No immediate action/update is required right now, we are first gathering opinions from all authors on this.

Thanks a lot for your contribution, and please ping me if you have any questions. If you have technical questions, also feel free to ping @kotekjedi who is happy to help with implementation details.

**Comment by stes** [comment]
Hi!

This is an automated report for the documentation coverage of your PR. To provide good user experience for the benchmark, we aim to cover most functions with doc strings, using the [Google docstring format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).

We would greatly appreciate your help in assuring that docstring coverage is sufficiently high for your contribution.

The current coverage is:
| Name                                         | Total | Miss | Cover | Cover% |
|----------------------------------------------|-------|------|-------|--------|
| examples/__init__.py                         |     1 |    1 |     0 |     0% |
| examples/imagenet_d.py                       |     9 |    9 |     0 |     0% |
| examples/imagenet_d_helper/map_classes.py    |     8 |    8 |     0 |     0% |
| shifthappens/models/torchvision.py           |     4 |    0 |     4 |   100% |
|----------------------------------------------|-------|------|-------|--------|
| TOTAL                                        |    22 |   18 |     4 |  18.2% |
RESULT: FAILED (minimum: 80.0%, actual: 18.2%)

You can reproduce this result by running

```
interrogate -v --ignore-nested-functions --ignore-private --ignore-semiprivate --ignore-init-method
```

on the files you added to your PR.

File: examples/imagenet_d_helper/map_classes.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (georgepachitariu) [reply examples/imagenet_d_helper/map_classes.py:89]: got it. I moved it.

File: examples/imagenet_d_helper/map_classes.py (hunk 1 above)
Comment (georgepachitariu) [reply examples/imagenet_d_helper/map_classes.py:390]: Oh, I got it now. Thank you :)

**Comment by georgepachitariu** [comment]
Hi @stes, 
My idea for integration test is to have a small archive that contains 1% of the data. It is put online and we run it as an extra task.

Initially, I was thinking of having this 1% of the data as part of the repository, but the problem (I think) is that the code expects the folders from all classes to be present and for imagenet-d/clipart that means adding 345 folders.

**Comment by stes** [comment]
> Hi @stes, My idea for integration test is to have a small archive that contains 1% of the data. It is put online and we run it as an extra task.
> 
> Initially, I was thinking of having this 1% of the data as part of the repository, but the problem (I think) is that the code expects the folders from all classes to be present and for imagenet-d/clipart that means adding 345 folders.

Thanks @georgepachitariu . We are discussing internally and will get back to you once we reached a conclusion. (e.g. to do what you proposed automatically on all tasks)

File: shifthappens/tasks/imagenet_d/imagenet_d.py [hunk 5]
```diff
@@ -24,16 +24,16 @@
 
 
 @dataclasses.dataclass
-class ImageNetSingleCorruptionTypeBase(Task):
+class ImageNetDBase(Task):
```
[the remark this answers is not in the document]
Comment (georgepachitariu) [reply shifthappens/tasks/imagenet_d/imagenet_d.py]: sure, no problem with that.

**Review by zimmerrol — APPROVED** [review]

**Pull request merged**